Readme.md
# Welcome to the ACE-BOX

The ace-box is an all-in-one Autonomous Cloud Enablement machine that you can use as a portable sandbox, demo and testing environment. 

## Check out [Troubleshooting](#troubleshooting) before reaching out!

Vagrant (Local) or Terraform (Cloud) are used for spinning up the VM, Ansible is used for setting up the various components.
- [Welcome to the ACE-BOX](#welcome-to-the-ace-box)
  - [Check out Troubleshooting before reaching out!](#check-out-troubleshooting-before-reaching-out)
  - [Release notes](#release-notes)
  - [Deployment Modes](#deployment-modes)
  - [Components](#components)
  - [Installation](#installation)
    - [Useful Terraform Commands](#useful-terraform-commands)
  - [Alt: Local installation with Vagrant](#alt-local-installation-with-vagrant)
    - [SSH into the box](#ssh-into-the-box)
    - [Vagrant cleanup](#vagrant-cleanup)
  - [Alt: Bring-your-own-VM](#alt-bring-your-own-vm)
  - [Default mode](#default-mode)
    - [Configuration settings](#configuration-settings)
    - [Resource Requirements](#resource-requirements)
  - [Troubleshooting](#troubleshooting)
  - [Accessing ACE Dashboard](#accessing-ace-dashboard)
  - [Behind the scenes](#behind-the-scenes)
  - [ACE-CLI](#ace-cli)
    - [Available commands (ace-cli version 0.0.1, can also be retrieved by running `ace --help`):](#available-commands-ace-cli-version-001-can-also-be-retrieved-by-running-ace---help)
    - [Available install components:](#available-install-components)
    - [Available uninstall components:](#available-uninstall-components)


## Release notes
Please check [RELEASE_NOTES.md](RELEASE_NOTES.md)

## Deployment Modes
The ACE-BOX comes in two deployment modes based on your requirement:

  1. Local installation (on your workstation) using vagrant
  2. Cloud installation using terraform

## Components
ACE-BOX comes with the following components

| Component | Version |
|----|-------|
| microk8s | 1.23 |
| jenkins | lts (2.319.1-lts at time of writing) |
| helm | 3 |
| oneagent | latest |
| activegate for private synthetic node | latest |
| ace dashboard | built on the spot |
| gitea local git server | 1.14.6 |
| keptn | 0.11.2 |


## Installation
The recommended way of installing any ACE box version, local or cloud, is via Terraform (scroll down for alternatives). Check the [Azure](terraform/azure/Readme.md), [AWS](terraform/aws/Readme.md), [Google Cloud](terraform/gcloud/Readme.md) and [Vagrant](terraform/vagrant/Readme.md) subfolders for additional instructions.

1. Check prereqs:
     - Terraform installed (+ Vagrant for local installation)
     - Dynatrace tenant (prod or sprint, dev not recommended)
2. Go to folder `./terraform/<aws, azure, gcloud or vagrant>/`
3. Set required Terraform variables:
   1. By adding `dt_tenant, dt_api_token, dt_paas_token` to a `terraform.tfvars` file:
      ```
      dt_tenant = "https://....dynatrace.com"
      dt_api_token = "dt0c01...."
      dt_paas_token = "dt0c01...."
      ```
   2. By setting environment variables:
      ```
      export TF_VAR_dt_tenant=https://....dynatrace.com
      export TF_VAR_dt_api_token=dt0c01....
      export TF_VAR_dt_paas_token=dt0c01....
      ```
    For details and alternatives see https://www.terraform.io/docs/language/values/variables.html

4. Run `terraform init`
5. Run `terraform apply`
6. Grab a coffee, this process will take some time...


### Useful Terraform Commands

Command  | Result
-------- | -------
`terraform destroy` | deletes any resources created by Terraform |
`terraform plan -destroy` | view a speculative destroy plan, to see what the effect of destroying would be |
`terraform show` | Outputs the resources created by Terraform. Useful to verify IP addresses and the dashboard URL. 


## Alt: Local installation with Vagrant
The local version can also be installed via Vagrant without the need for Terraform:

1. Check prereqs:
    - A workstation with at least **16GB of RAM** and **4 CPU cores (non-virtualized)**
    - Virtualbox installed (6.1.x tested)
    - Vagrant installed (2.2.7 tested)
    - Dynatrace tenant (prod or sprint, dev not recommended)
    - Hyper-V disabled on Windows machines, check [Troubleshooting](#troubleshooting) for more information
1. Go to folder `./terraform/vagrant/`
2. Create `ace.config.yml` e.g. by renaming `ace.config.yml.tpl` in place or copying from `refs`)
3. Set required variables:
    ```
    ---
    dynatrace:
      tenant:     "https://....dynatrace.com"
      apitoken:   "dt0c01...."
      paastoken:  "dt0c01...."
    ...
    ```
4. Run `vagrant up`
5. Grab a coffee, this process will take some time...

**Note:** The first time you might need to enter your passord at least once.

**Note:** Windows users will be asked to confirm security notifications a couple of times during the provisioning process, so keep an eye out for them.

### SSH into the box
Execute `vagrant ssh` to gain access to the VM

### Vagrant cleanup

Vagrant offers many commands to deal with the VM, check the below:

Command  | Result
-------- | -------
`vagrant destroy` | stops and deletes all traces of the vagrant machine |
`vagrant halt` | stops the vagrant machine - i.e. shutting down your workstation |
`vagrant suspend` | suspends the machine - i.e. sleep your workstation |
`vagrant resume` | resume a suspended vagrant machine |
`vagrant up` | starts and provisions the vagrant environment |
`vagrant box update` | update the base box from time to time to ensure it is the latest version. While provisioning a message will be shown that there are updates available |


## Alt: Bring-your-own-VM

Bringing your own Ubuntu Virtual Machine has not been tested, but should be possible:

1. Check prereqs:
    - An Ubuntu 18.04 virtual machine (Ubuntu 18.04 LTS "minimal" tested)
    - A public IP address
    - Port 80 and/or 443 exposed
    - A non-root user to run the script actions needs to be created (e.g. `ace`)
    - Repository cloned to VM
2. Run initialization script:
    ```
    $ cd user-skel
    $ ./init.sh
    ```
  This will install all necessary dependencies including the ace-cli.

3. Prepare ACE-Box by running the ace-cli and providing required values when prompted:
      ```
      $ ace prepare
      ```

4. Install ACE-Box components:
      ```
      $ ace install default
      ```
5. Grab a coffee, this process will take some time...


## Default mode

Below config example will spin up an ace-box with all features in a demo mode (everything pre-configured). In demo mode, the following gets installed and configured:
- Microk8s 
- Helm 
- Dynatrace Acitve Gate for Private Synthetic gets installed and automatically configured in the Dynatrace tenant as a Private Location
- Dynatrace OneAgent operator gets installed via Helm
- Gitea gets installed, a user gets created and a repository created
- Keptn Quality Gates gets installed
- Jenkins gets installed with all the plugins that are needed, pipeline libraries get added for interaction with keptn and Dynatrace, pipelines get created and linked to the gitea repository
- A dashboard gets built and deployed with links to all the components as well as credentials
- A kubernetes ingress gets configured for all applications so it is easy to navigate
- Dynatrace tag rules and request attributes get set up via the API


### Configuration settings

The ace-box comes with a certain number of features and settings that can be set/enabled/disabled. Adding and removing features will change the resource consumption. Most settings have default values and do not need to be set explicitly, but they can be overwritten if needed. Please refer to the ace cli instruction below.


### Resource Requirements
Each feature requires a certain amount of resources - on top of the base microk8s requirements.
The resource requirements below are measured using Dynatrace's Kubernetes monitoring
Feature  | Kubernetes Resource Usage | 
-------- | ------- |
GitLab | 11 mCores, 2GB RAM
Jenkins | 1mCore, 1GB RAM
Gitea | 2 mCores, 250MB RAM


## Troubleshooting
1. During testing it was found that when spinning up the VM while being connected to the corporate VPN it would sometimes have connectivity issues. It is best to disconnect from the VPN while provisioning. This will also drastically speed up the provision process. VPN issues manifests themselves mainly in Jenkins being empty (no pipelines or plugins installed) after provisioning. If you have this, turn off VPN and re-provision.
2. Some users had issues with (old) customer vpn software that was installed - not even connected -  causing issues with the virtual network adaptors. If you are having issues provisioning the VM, uninstall them when possible
3. If you are using a Windows workstation, ensure that Hyper-V native virtualization has been disabled as it clashes with virtualbox. Hyper-V support is on the roadmap. Check this [doc](https://docs.microsoft.com/en-us/virtualization/hyper-v-on-windows/quick-start/enable-hyper-v) on how to disable Hyper-V
4. If at any given time the provisioning fails, it is best to execute a `vagrant destroy` followed by a `vagrant up`
5. During testing there were some cases where Jenkins plugins refused to install while provisioning which renders the installation useless for the other usecases. In that case, it is best to execute a `vagrant destroy` followed by a `vagrant up`
6. On Windows machines the following error might occur `Stderr: VBoxManage.exe: error: Failed to open/create the internal network 'HostInterfaceNetworking-VirtualBox Host-Only Ethernet Adapter #2' (VERR_INTNET_FLT_IF_NOT_FOUND)`. It was found that disabling and enabling the network adaptor solved the issue
7. On Windows machines it might be that a `vagrant ssh` `gives vagrant@127.0.0.1: Permission denied (publickey,gssapi-keyex,gssapi-with-mic)` or similar. A common cause is that a privatekey file has too many people with access. This was seen when the repo was cloned in a subfolder of the C drive on Dynatrace laptops. It is suggested to use a subfolder of your home directory.
8. Your DNS settings might hinder the provisioning of the ace-box as the name does not resolve. It mainly manifests itself when it is waiting for gitea to be up. If you get a message like the below, you are most likely affected. The best way to go around it, is by changing the DNS settings on your network adaptor and point to for example the google DNS servers (8.8.8.8 and 8.8.4.4). A quick google search should tell you how to do that for your particular OS.
    ```
    FAILED - RETRYING: Gitea - Wait for API to be up (1 retries left).
    fatal: [ace-box]: FAILED! => {"attempts": 60, "changed": false, "content": "", "elapsed": 0, "msg": "Status code was -1 and not [200]: Request failed: <urlopen error [Errno -5] No address associated with hostname>", "redirected": false, "status": -1, "url": "http://gitea.192.168.50.10.nip.io/api/v1/admin/orgs?access_token=1c8d4fcef25b3ae2a15d17d29be64c2c7aa22501"}
    ```
9. Dynatrace Operator installation fails with "Error: Cluster already exists: ...": If you ever had a cluster created before please remove it from https://<dynatrace tenant>/#settings/kubernetesmonitoring;gf=all

## Accessing ACE Dashboard
At the end of the provisioning, an ACE Dashboard gets created with more information on how to use the ACE-BOX. Check out [ACE Dashboard](Dashboard.md) for more details.

## Behind the scenes
Spinning up an ACE-Box can be split into two main parts:

1) Deploying a VM: This can either happen locally via Vagrant or a VM is can be created in your Cloud Account.
2) After a VM is available, provisioners install the actual application (i.e. "ACE-Box" logic). This process itself consists of a couple steps:
   1) Copying working directory: Everything in [user-skel](/user-skel) is copied to the VM
   2) Package manager update: [init.sh](/user-skel/init.sh) is run. This runs an `apt-get` update and installs `Python3.9`, Ansible and the `ace-cli`
   3) `ace prepare` is run, which asks for ACE-Box specific configurations (e.g. protocol, custom domain, ...)
   4) Once the VM is prepared, the actual installation happens by running `ace install default`

## ACE-CLI
`ACE-Box` comes with an including management tool called 'ace-cli'. This cli tool can be used to prepare and/or install the ACE-Box or certain components.

```
$ ace --version
```

### Available commands (ace-cli version 0.0.1, can also be retrieved by running `ace --help`):

  Command | Result |
  -- | -- |
  `prepare` | Prepares ACE-Box for further use (e.g. persists domain, protocol settings) |
  `install <component>` | Installs ACE-Box or components thereof (see table below) |
  `uninstall <component>` | Uninstalls ACE-Box or components thereof (see table below) |
  `set <config>` | Updates ACE-Box config. <config> can be any of key=value, e.g. `$ace set feature_gitlab=true` |

### Available install components:

  Component | Result |
  -- | -- |
  `default` | Installs and configures everything needed for a quality gates demo (formerly known as "demo" mode) |
  `microk8s` | Installs and configures MicroK8S |
  `gitea` | Installs and configures Gitea |
  `gitlab` | Installs and configures Gitlab |
  `dynatrace` | Installs and configures Dynatrace OneAgent and ActiveGate |
  `repositories` | Installs and configures Git Repositories |
  `keptn` | Installs and configures Keptn |
  `monaco` | Installs and configures Monaco |
  `jenkins` | Installs and configures Jenkins |
  `dashboard` | Installs and configures the dashboard |
  `awx` | Installs AWX |


### Available uninstall components:

Please refer to `$ ace --help` for available commands


RELEASE_NOTES.md
# ACE-BOX Release notes

## v1.10.0 - January 2022

### New Features

#### AppSec Quality Gate / Security Gate
A new pipeline set has been integrate to demonstrate AppSec Quality Gates!
[#116](https://github.com/Dynatrace/ace-box/issues/116)

#### Provisioning improvements
Improvements in provisioning and updating ace-box instances.
[#58](https://github.com/Dynatrace/ace-box/issues/58)

In a nutshell:

- Unified provisioning: Terraform for cloud + Vagrant instances (Vagrant legacy can be used as well)
- Ansible cleanup: Move playbooks and related config (e.g. variables) into roles
- Installation: install.sh is split into:
   - init.sh: runs some essentials (e.g. apt-get update, install Ansible, ... + "ace-cli"). This usually needs to be run once for every ACE-box and is included in the Terraform provisioners.
   - "$ ace prepare": Adds VM specific settings (e.g. ingress domain). This usually needs to be run once for every ACE-box and is included in the Terraform provisioners.
   - "$ ace install ": Installs ACE-Box components, e.g. "all", "microk8s", ...
Please see the README for additional info.

#### Documentation updates
monaco-gitops-demo usecase documentation
[#95](https://github.com/Dynatrace/ace-box/issues/95)

First build might see quality gate evaluation fail due to tag propagation timing issue
[#96](https://github.com/Dynatrace/ace-box/issues/96)

### Fixes/tech updates
Stability issues with longer running ace-box instances fixed by updating microk8s and installing `iptables-persistent` package
[#124](https://github.com/Dynatrace/ace-box/issues/124)

Keptn update to 0.11.2
[#109](https://github.com/Dynatrace/ace-box/issues/109)

Jenkins update to 2.319.1-lets
[#120](https://github.com/Dynatrace/ace-box/issues/120)


## v1.9.5 - November 2021

There are no new features in this release, only targetting fixes and imporovements
### Fixes/tech updates

#### Keptn 0.10.0 Update and support proper merging of metrics
Update to latest version of Keptn and Dynatrace Service for Keptn
[#86](https://github.com/Dynatrace/ace-box/issues/86)
[#81](https://github.com/Dynatrace/ace-box/issues/81)
#### Dynatrace Operator
The Operator install script was intermittently failing. Went back to manual install.
[#70](https://github.com/Dynatrace/ace-box/issues/70)

#### GCP Terraform requires sourcerange firewall
Library update
[#88](https://github.com/Dynatrace/ace-box/pull/88)

#### Dashboard deployed from image registry
Speeds up deployment
[#32](https://github.com/Dynatrace/ace-box/discussions/32)
#### Various
Monaco failure 
[#77](https://github.com/Dynatrace/ace-box/issues/77)

Gitea login enforced 
[#60](https://github.com/Dynatrace/ace-box/issues/60)

## v1.9 - July 2021

### New Features

#### Custom domains support
[#3](https://github.com/Dynatrace/ace-box/issues/3)

This allows you to use custom domains, either directly in terraform or by running the install script on a VM.

#### Canary Release and Auto Remediation usecase
[#11](https://github.com/Dynatrace/ace-box/issues/11) and [#20](https://github.com/Dynatrace/ace-box/issues/20)

You can now demo a canary deployment together with auto remediation using the ACE-BOX. This will use Jenkins, Nginx Ingress and AWX to drive the workflow. Check out the new usecase from the ACE Dashboard!
> Note: this usecase works only on a cloud ace-box as Dynatrace has to be able to send the problem back to AWX

#### AWX
[#11](https://github.com/Dynatrace/ace-box/issues/11) and [#20](https://github.com/Dynatrace/ace-box/issues/20)

AWX Is used by the Auto Remediation usecase and thus also gets installed on the ACE-BOX.

#### Complete demo walkthrough
[#29](https://github.com/Dynatrace/ace-box/issues/29)

On the ACE Dashboard there is now a detailed walkthrough for the demo scenarios of MaaS/Quality Gates and also Canary/Auto Remediation.

### Fixes/tech updates

#### K8s 1.21
[#38](https://github.com/Dynatrace/ace-box/issues/38)

Microk8s has been upgraded from 1.18/stable to 1.21/stable. This was done for tech currency reasons and also to support the canary scenario using the nginx ingress controller

#### Update AWS Instance Type to T3
[#15](https://github.com/Dynatrace/ace-box/issues/15)

#### Added CODEOWNERS
[#21](https://github.com/Dynatrace/ace-box/issues/21)

## v1.8 - June 2021

This is the last version of the ACE-BOX to be published in BitBucket. From now on, the repository will be maintained in https://github.com/Dynatrace/ace-box. You need a GitHub account linked to the Dynatrace org to gain access. Please lodge an EDE ticket.

### New features

#### Azure and AWS support for Cloud Edition
Following on the first release of ACE-BOX CE, support for Azure and AWS using terraform has now been added. Check out the **microk8s/terraform** folder for more info. Terraform config has also been restructured.

#### Basic authentication for ACE Dashboard
Since the dashboard prints user names and tokens for gitea and other services, basic authentication has been added. User name and password are set to `dynatrace` by default but can be changed by setting the `acebox.config.dashboard.user` and `acebox.config.dashboard.password` in `config.yml`.

#### A brand new ACE Dashboard experience
We moved the ACE Dashboard to the barista framework to be in line with the Dynatrace design. At the same time, the goal is to make the ACE-BOX more approachable so we have made it even easier to get started!

#### Clean up step for Dynatrace Environment prior to deployment
An extra step has been added that deletes config from Dynatrace. At the moment the Kubernetes credentials for `127.0.0.1` and `ACE-BOX` are deleted to avoid conflicts with the new Dynatrace Operator and using a Dynatrace environment that was used prior for the ace-box.

#### GitLab CI Demo
By importing the ace/ace repo from Gitea into the (optional) GitLab environment, you can now showcase Quality Gates and event from within a GitLab environment.

### Fixes/tech updates

microk8s/repos/hot-repo/app.js
const EMPTY = "<EMPTY>";
var port = process.env.PORT || 8080,
    http = require('http'),
    fs = require('fs'),
	os = require('os'),
	sem = require('semver'),
	dttags = process.env.DT_TAGS || EMPTY,
	dtcustprops = process.env.DT_CUSTOM_PROP || EMPTY,
	dtclusterid = process.env.DT_CLUSTER_ID || EMPTY,
	namespace = process.env.NAMESPACE || EMPTY,
	pod_name = process.env.POD_NAME || EMPTY,
	deployment_name = process.env.DEPLOYMENT_NAME || EMPTY,
	container_image = process.env.CONTAINER_IMAGE || EMPTY,
	keptn_project = process.env.KEPTN_PROJECT || EMPTY,
	keptn_stage = process.env.KEPTN_STAGE || EMPTY,
	keptn_service = process.env.KEPTN_SERVICE || EMPTY,
    html = fs.readFileSync('index.html').toString().replace("HOSTNAME", os.hostname()); //  + " with DT_TAGS=" + dttags + "\nDT_CUSTOM_PROP=" + dtcustprops + "\nDT_CLUSTER_ID=" + dtclusterid);
	sver = sem.valid('1.2.3')
// semver is loaded for a medium level appsec vulneraility
// ======================================================================
// Here are some global config entries that change the behavior of the app
// ======================================================================
var buildNumber = 1;
var minSleep = 200;
var requestCount = 0;
var inProduction = false;
var invokeRequestCount = 0;
var failInvokeRequestPercentage = 0;


// collect request info
var requests = [];
var requestTrimThreshold = 5000;
var requestTrimSize = 4000;


// ======================================================================
// does some init checks and sets variables!
// ======================================================================
var init = function(newBuildNumber) {
	// MAKE SURE we have a good NAMESPACE
	if(!namespace || (namespace.length == 0) || (namespace == EMPTY)) {
		if(keptn_stage && keptn_stage.length)
			namespace = keptn_stage;	
		else if(deployment_name && deployment_name.length)
			namespace = deployment_name;	
		else if(keptn_stage && pod_name.length)
			namespace = pod_name;	
	}

	// CHECK IF WE ARE RUNNING "In Production"
	// first we check if somebody set the deployment_group_name env-variable
	inProduction = process.env.DEPLOYMENT_GROUP_NAME && process.env.DEPLOYMENT_GROUP_NAME.startsWith("Production");
	// second we check whether our host or podname includes blue or green in its name - we use this for blue/green deployments in production
	if(!inProduction) {
		inProduction = os.hostname().includes("green") || os.hostname().includes("blue");
		if(!inProduction) {
			inProduction = namespace && namespace.toLowerCase().includes("prod");
		}
	}
	
	if(inProduction) {
		minSleep = 100; // we just simulate that production is a bit faster than staging, e.g: better hardware!
	}

	// here are some "problems" we simulate for different builds. Builds are identified via Env Variable BUILD_NUMBER;
	// Build # | Problem
	// 1 | no problem
	// 2 | 50% of requests return HTTP 500 Status Code
	// 3 | back to normal
	// 4 | no problem in staging but problem in prod -> higher sleep time and 10% of requests fail
	// X | any other build number will run like 1 & 3
	if(newBuildNumber != null) {
		buildNumber = parseInt(newBuildNumber);
	}
	else if(process.env.BUILD_NUMBER && process.env.BUILD_NUMBER != null) {
		buildNumber = parseInt(process.env.BUILD_NUMBER);
    }

	switch(buildNumber) {
		case 2:
			//failInvokeRequestPercentage = 50;
			minSleep = 600;
			break;
		case 4: 
			minSleep = minSleep * 2;
			if(inProduction) {
				failInvokeRequestPercentage = 20;
			}
			break;
		case 5:
			// introduce appsec issue
			var merge = require("@brikcss/merge")
			var obj = {}
			var malicious_payload = '{"__proto__":{"polluted":"Yes! Its Polluted"}}';
			console.log("Before: " + {}.polluted);
			merge({}, JSON.parse(malicious_payload));
			console.log("After : " + {}.polluted);
			break;
		default:
			// everything normal here
			failInvokeRequestPercentage = 0;		
			break;
	}

	console.log("Init: " + buildNumber + "/" + failInvokeRequestPercentage + "/" + minSleep);
} 

// ======================================================================
// Background colors for our app depending on the build
// ======================================================================
var backgroundColors = ["#D6D4D2", "#73A53E", "#FF7C00", "#D3D309", "#4AB9D9"]
var getBackgroundColor = function() {
	var buildNumberForBackgroundColor = buildNumber;
	if(buildNumber == 0 || buildNumber > 4) buildNumberForBackgroundColor = 1;
	
	return backgroundColors[buildNumberForBackgroundColor];
}


// ======================================================================
// This is for logging
// ======================================================================
//var logstream = fs.createWriteStream('./serviceoutput.log');
var SEVERITY_DEBUG = "Debug";
var SEVERITY_INFO = "Info";
var SEVERITY_WARNING = "Warning";
var SEVERITY_ERROR = "Error";

var log = function(severity, entry) {
	if (severity === SEVERITY_DEBUG) {
		// dont log debug
	} else {
		var logEntry = new Date().toISOString() + ' - ' + severity + " - " + entry + '\n';
		//logstream.write(logEntry);
		//console.log is better suited for containers
		console.log(logEntry);
	}
};

// ======================================================================
// Very inefficient way to "sleep"
// ======================================================================
function sleep(time) {
	if(time < minSleep) time = minSleep;
	var stop = new Date().getTime();
    while(new Date().getTime() < stop + time) {
		;
    }
}

function getRequestsPerMinute() {
	var now = Date.now();
	var aMinuteAgo = now - (1000 * 60);
	var cnt = 0;
	// since recent requests are at the end of the array, search the array
	// from back to front
	for (var i = requests.length - 1; i >= 0; i--) {
		if (requests[i] >= aMinuteAgo) {
			++cnt;
		} else {
			break;
		}
	}
	return cnt
}

// ======================================================================
// This is our main HttpServer Handler
// ======================================================================
var server = http.createServer(async function (req, res) {

	requests.push(Date.now());

	// now keep requests array from growing forever
	if (requests.length > requestTrimThreshold) {
		requests = requests.slice(0, requests.length - requestTrimSize);
	}
	log(SEVERITY_INFO, req.method + ' - ' + req.url);

    if (req.method === 'POST') {
        var body = '';

        req.on('data', function(chunk) {
            body += chunk;
        });

        req.on('end', function() {
            if (req.url === '/') {
                log(SEVERITY_DEBUG, 'Received message: ' + body);
            } else if (req.url = '/scheduled') {
                log(SEVERITY_DEBUG, 'Received task ' + req.headers['x-aws-sqsd-taskname'] + ' scheduled at ' + req.headers['x-aws-sqsd-scheduled-at']);
            }

            res.writeHead(200, 'OK', {'Content-Type': 'text/plain'});
            res.end();
        });
    } else if (req.url.startsWith("/api")) {
		var url = require('url').parse(req.url, true);
		var closeResponse = true;

        // sleep a bit :-)
		var sleeptime = parseInt(url.query["sleep"]);
		if(sleeptime === 0 || isNaN(sleeptime) ) sleeptime = minSleep;
		log(SEVERITY_INFO, "Sleeptime: " + sleeptime);
		sleep(sleeptime);

		// figure out which API call they want to execute
        var status = "Unkown API Call";
		if(url.pathname === "/api/sleeptime") {
			// Usage: /api/sleeptime?min=1234 
			var sleepValue = parseInt(url.query["min"]);
			if(sleepValue >= 0 && sleepValue <= 10000) minSleep = sleepValue;
			status = "Minimum Sleep Time set to " + minSleep;
		}
		if(url.pathname === "/api/echo") {
			// Usage: /api/echo?text=your text to be echoed!
			status = "Thanks for saying: " + url.query["text"];
		}
		if(url.pathname === "/api/login") {
			// Usage: /api/login?username=your user name 
			status = "Welcome " + url.query["username"];
		}
		if(url.pathname === "/api/invoke") {
			// count the invokes for failed requests
			var returnStatusCode = 200;
			if(failInvokeRequestPercentage > 0) {
				invokeRequestCount++;
				var failRequest = (invokeRequestCount % (100 / failInvokeRequestPercentage));
				console.log(invokeRequestCount + "%" + failInvokeRequestPercentage + "=" + failRequest);
				if(failRequest == 0) {
					returnStatusCode = 500;
					invokeRequestCount = 0;
				}
			}

			// Usage: /api/invoke?url=http://www.yourdomain.com 
			var urlRequest = url.query["url"];
			status = "Trying to invoke remote call to: " + urlRequest;
			
			var http = null;
			if(urlRequest.startsWith("https")) http = require("https");
			else http = require("http");
			closeResponse = false;
			var options = {
              	host: urlRequest,
              	path: '/'
            };
			var result = http.get(urlRequest, function(getResponse) {
				log(SEVERITY_DEBUG, 'STATUS: ' + getResponse.statusCode);
				log(SEVERITY_DEBUG, 'HEADERS: ' + JSON.stringify(getResponse.headers));

				// Buffer the body entirely for processing as a whole.
				var bodyChunks = [];
				getResponse.on('data', function(chunk) {
					bodyChunks.push(chunk);
				}).on('end', function() {
					var body = Buffer.concat(bodyChunks);
				  	log(SEVERITY_DEBUG, 'BODY: ' + body);
				  	status = "Request to '" + url.query["url"] + "' returned with HTTP Status: " + getResponse.statusCode + " and response body length: " + body.length;
				  	res.writeHead(returnStatusCode, returnStatusCode == 200 ? 'OK' : 'ERROR', {'Content-Type': 'text/plain'});	
				  	res.write(status);
				  	res.end();
				}).on('error', function(error) {
				  	status = "Request to '" + url.query["url"] + "' returned in an error: " + error;
				  	res.writeHead(returnStatusCode, returnStatusCode == 200 ? 'OK' : 'ERROR', {'Content-Type': 'text/plain'});	
				  	res.write(status);
				  	res.end();					
				  	log(SEVERITY_INFO, status);
				})
			});
		}
		if(url.pathname === "/api/version") {
			if (url.query["newBuildNumber"] && url.query["newBuildNumber"] != null) {
				var newBuildNumber = url.query["newBuildNumber"];
				log(SEVERITY_WARNING, "Somebody is changing! buildNumber from " + buildNumber + " to " + newBuildNumber);

				init(newBuildNumber);
			}

			// usage: /api/version
			// simply returns the build number as defined in BUILD_NUMBER env-variable which is specified
			status = "Running build number: " + buildNumber + " Production-Mode: " + inProduction;
			status += "\n\nHere some additional environment variables:";
			status += "\nKEPT_PROJECT: " + keptn_project;
			status += "\nKEPTN_STAGE: " + keptn_stage;
			status += "\nKEPTN_SERVICE: " + keptn_service;
			status += "\nDT_TAGS: " + dttags;
			status += "\nDT_CUSTOM_PROP: " + dtcustprops;
			status += "\nDT_CLUSTER_ID: " + dtclusterid;
			status += "\nDEPLOYMENT_NAME: " + deployment_name;
			status += "\nCONTAINER_IMAGE: " + container_image;
			status += "\nPOD_NAME: " + pod_name;
			status += "\nNAMESPACE: " + namespace;
		}
		if(url.pathname === "/api/causeerror") {
			log(SEVERITY_ERROR, "somebody called /api/causeerror");
			status = "We just caused an error log entry"
		}
		if (url.pathname === "/api/cpuload") {
			const reqPerMin = getRequestsPerMinute();
			let sleepTime = 1200;

			if (reqPerMin <= 70) {
				sleepTime = Math.pow(reqPerMin, 2) - Math.pow(reqPerMin, 3) / 100;
			}

			if (reqPerMin <= 45) {
				sleepTime = (Math.pow(reqPerMin, 2) - Math.pow(reqPerMin, 3) / 100) / 2;
			}

			console.log("Sleeping for " + sleepTime + "ms");
			sleep(sleepTime);

			status = "Request finished";
		}

		// only close response handler if we are done with work!
		if(closeResponse) {
		   res.writeHead(200, 'OK', {'Content-Type': 'text/plain'});	
		   res.write(status);
		   res.end();
		}
	}
	else
	{
		res.writeHead(200, 'OK', {'Content-Type': 'text/html'});

		// replace buildnumber and background color
		var finalHtml = html.replace("BACKGROUND-COLOR", getBackgroundColor()).replace("BUILD_NUMBER", buildNumber).replace("NAMESPACE", namespace);
        res.write(finalHtml);
        res.end();
	}
	
	requestCount++;
	if(requestCount >= 100) {
		log(SEVERITY_INFO, "Just served another 100 requests!");
		requestCount = 0;
	}
});

// first we initialize!
init(null);

// Listen on port 80, IP defaults to 127.0.0.1
server.listen(port);

// Put a friendly message on the terminal
console.log('Server running at http://127.0.0.1:' + port + '/');
log(SEVERITY_INFO, "Service is up and running - feed me with data!");